### PR TITLE
Add notification system

### PR DESCRIPTION
**Pull request by Yassa-hue**
## The notifications system
- Implement the notification page.
- Add the notification model.
- Add the notification service.
- Send a notification to all the donors and volunteers when a new campaign is added.
- Send a notification to the target organization when a donor donates or a volunteer requests to join.

### CODE REVIEW

File: lib/services/donation_service.dart [hunk 1]
```diff
@@ -10,12 +12,21 @@ class DonationService {
...
     await docRef.update({'id': donationId});
+
+    NotificationModel notification = NotificationModel(
+      title: 'New Donation',
+      body:
+          'A new donation has been made to your campaign: ${donation.amount}\$',
+      userId: donation.receivingOrganizationId,
+    );
+
+    await NotificationService().createNotification(notification);
```
Comment (ihab4real) [inline lib/services/donation_service.dart:23]: let's send this donation notification back to the donor of the donation as well.

**Review by ihab4real — COMMENTED** [review]
these were some good stuff

File: lib/services/notification_service.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    if (snapshot.exists) {
+      return NotificationModel.fromJson(snapshot.data() as Map<String, dynamic>);
+    } else {
+      throw Exception('Notification not found');
+    }
+  }
+
+  Future<void> updateNotification(NotificationModel notification) async {
+    await _colectionReference.doc(notification.id).update(notification.toJson());
+  }
```
[the remark this answers is not in the document]
Comment (Yassa-hue) [reply lib/services/notification_service.dart:28]: There is no use case for that for now. But a service should have all the CRUD operations anyway.

File: lib/services/notification_service.dart [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  Future<void> deleteNotification(String id) async {
+    await _colectionReference.doc(id).delete();
+  }
+
+  Future<List<NotificationModel>>? getNotificationsByUser(String userId) async {
```
[the remark this answers is not in the document]
Comment (Yassa-hue) [reply lib/services/notification_service.dart:34]: We use this notation all over the code base, the method should take a user model object not a userId in theory. ByUserId is too specific, we want to get the notification of a user.

File: lib/services/campaign_service.dart [hunk 4]
```diff
@@ -3,16 +3,33 @@ import 'dart:io';
...
+      title: 'New Campaign',
+      body: 'A new campaign has been created: ${campaign.title}',
+      userId: '',
+    );
+
+    await NotificationService()
+        .sendNotificationToUsersByType(notification, UserRole.donor);
+
+    await NotificationService()
+        .sendNotificationToUsersByType(notification, UserRole.volunteer);
```
[the remark this answers is not in the document]
Comment (Yassa-hue) [reply lib/services/campaign_service.dart:31]: Ideally, we should have a matching algorithm to select only the users (volunteers and donors) who may be interested in this event based on their profile. Now we don't have such algo but we have a very small set of users so sending a notification to all of them won't be that expensive.

File: lib/services/donation_service.dart (hunk 1 above)
Comment (Yassa-hue) [reply lib/services/donation_service.dart:23]: Done!

**Pull request merged**